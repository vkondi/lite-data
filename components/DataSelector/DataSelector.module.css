.container {
    flex: 1;
}

.layoutContainer {
    display: flex;
    flex-direction: column;
}