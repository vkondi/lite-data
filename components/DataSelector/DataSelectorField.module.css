.container {
    display: flex;
    flex-direction: row;
    padding: 10px 5px;
    justify-content: space-between;
    column-gap: 10px;
}