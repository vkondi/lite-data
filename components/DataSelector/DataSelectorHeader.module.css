.container {
    display: flex;
    flex-direction: row;
    padding: 5px 5px;
    justify-content: space-between;
    column-gap: 10px;
    border-bottom: 1px solid black;
}

.deleteIconSpacer {
    height: 40px;
    width: 28px;
}