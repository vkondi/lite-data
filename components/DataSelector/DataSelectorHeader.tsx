import React from 'react';
import { FormControl, InputLabel } from '@mui/material';
import styles from './DataSelectorHeader.module.css';

const DataSelectorHeader: React.FC = () => {
  return (
    <div className={styles.container}>
      <FormControl variant="outlined" sx={{ flex: 1, borderRight: '1px solid #e0e0e0' }}>
        <InputLabel>Data Type</InputLabel>
      </FormControl>
      <FormControl variant="outlined" sx={{ flex: 1,borderRight: '1px solid #e0e0e0' }}>
        <InputLabel>Field Name</InputLabel>
      </FormControl>
      <div className={styles.deleteIconSpacer} data-testid="delete-icon-spacer"></div>
    </div>
  );
};

export default DataSelectorHeader;