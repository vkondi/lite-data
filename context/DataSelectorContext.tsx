'use client';

import axios from "axios";
import {
  createContext,
  Dispatch,
  FC,
  PropsWithChildren,
  SetStateAction,
  useContext,
  useEffect,
  useState,
} from "react";
import { GET_CONFIG_URL } from "@/utils/URL";

export type FieldData = {
  dataType: string;
  name: string;
};

type DataSelectorContextProps = {
  fields: FieldData[];
  setFields: Dispatch<SetStateAction<FieldData[]>>;
  allowedDataTypes: string[];
};

const emptyField: FieldData = { dataType: "", name: "" };
const defaultFields: FieldData[] = [emptyField];

const DataSelectorContext = createContext<DataSelectorContextProps | undefined>(
  undefined
);

export const DataSelectorProvider: FC<PropsWithChildren> = ({
  children,
}): JSX.Element => {
  const [allowedDataTypes, setAllowedDataTypes] = useState<string[]>([]);
  const [fields, setFields] = useState<FieldData[]>(defaultFields);

  const fetchConfig = async () => {
      try {
        const response = await axios.get(GET_CONFIG_URL);
        setAllowedDataTypes(response?.data?.allowedDataTypes ?? []);
      } catch (error) {
        console.error("Failed to fetch config:", error);
      }
    };

  useEffect(() => {
    fetchConfig();
  }, []);

  return (
    <DataSelectorContext.Provider value={{ fields, setFields, allowedDataTypes }}>
      {children}
    </DataSelectorContext.Provider>
  );
};

export const useDataSelectorContext = (): DataSelectorContextProps => {
  const context = useContext(DataSelectorContext);

  if (!context) {
    throw new Error(
      "useDataSelectorContext must be used within a DataSelectorProvider"
    );
  }
  return context;
};
