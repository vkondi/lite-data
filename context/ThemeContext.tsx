'use client';

import React, { createContext, useContext, useState, useCallback, useEffect } from 'react';

type ThemeContextType = {
  mode: 'light' | 'dark';
  toggleColorMode: () => void;
};

const ThemeContext = createContext<ThemeContextType>({
  mode: 'light',
  toggleColorMode: () => {},
});

export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [mode, setMode] = useState<'light' | 'dark'>('light');

  useEffect(() => {
    localStorage.setItem('theme', mode);
  }, [mode]);

  const toggleColorMode = useCallback(() => {
    setMode((prevMode) => (prevMode === 'light' ? 'dark' : 'light'));
  }, []);

  return (
    <ThemeContext.Provider value={{ mode, toggleColorMode }}>
      {children}
    </ThemeContext.Provider>
  );
};

export const useThemeContext = () => {
  const context = useContext(ThemeContext);
  if (!context) {
    throw new Error('useThemeContext must be used within a ThemeProvider');
  }
  return context;
}; 