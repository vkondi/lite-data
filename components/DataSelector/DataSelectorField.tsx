/**
 * DataSelectorField component allows users to select a data type and provide a name for the field.
 * It also provides an option to delete the field.
 *
 * @component
 * @param {Object} props - The properties object.
 * @param {function} props.onTypeChange - Callback function to handle changes in the data type selection.
 * @param {function} props.onNameChange - Callback function to handle changes in the field name input.
 * @param {function} props.onDelete - Callback function to handle the deletion of the field.
 * @param {number} props.index - The index of the current field in the list.
 * @param {string} props.dataTypeValue - The current value of the data type selection.
 * @param {string} props.nameValue - The current value of the field name input.
 *
 * @returns {JSX.Element} The rendered DataSelectorField component.
 */

import React from "react";
import Select, { SelectChangeEvent } from "@mui/material/Select";
import MenuItem from "@mui/material/MenuItem";
import TextField from "@mui/material/TextField";
import IconButton from "@mui/material/IconButton";
import DeleteIcon from "@mui/icons-material/Delete";

import styles from "./DataSelectorField.module.css";
import FormControl from "@mui/material/FormControl";
import InputLabel from "@mui/material/InputLabel";
import { useDataSelectorContext } from "../../context/DataSelectorContext";

const DATA_TYPES = [
  { id: "name", label: "Name" },
  { id: "phone", label: "Phone" },
  { id: "number", label: "Number" },
  { id: "boolean", label: "Boolean" },
  { id: "date", label: "Date" },
  { id: "email", label: "Email" },
  { id: "time", label: "Time" },
  { id: "address", label: "Address" },
  { id: "city", label: "City" },
  { id: "country", label: "Country" },
  { id: "zipCode", label: "Zip Code" },
  { id: "company", label: "Company" },
  { id: "jobTitle", label: "Job Title" },
  { id: "color", label: "Color" },
  { id: "uuid", label: "UUID" },
  { id: "url", label: "URL" },
  { id: "ipAddress", label: "IP Address" },
  { id: "macAddress", label: "MAC Address" },
];

interface DataTypeFieldProps {
  onTypeChange: (value: string, index: number) => void;
  onNameChange: (value: string, index: number) => void;
  onDelete: (index: number) => void;
  index: number;
  dataTypeValue: string;
  nameValue: string;
}

const DataSelectorField: React.FC<DataTypeFieldProps> = ({
  onTypeChange,
  onNameChange,
  onDelete,
  index,
  dataTypeValue,
  nameValue,
}) => {
  const { allowedDataTypes } = useDataSelectorContext();
  
  const handleDelete = () => {
    onDelete(index);
  };

  const handleTypeChange = (event: SelectChangeEvent<string>) => {
    onTypeChange(event.target.value, index);
  };

  const handleNameChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;
    onNameChange(value, index);
  };

  const renderMenuItems = () => {
    return DATA_TYPES.map((dataType) => (
      <MenuItem
        key={dataType.id}
        value={dataType.id}
        disabled={!allowedDataTypes?.includes(dataType.id)}
      >
        {dataType.label}
      </MenuItem>
    ));
  };

  return (
    <div className={styles.container}>
      <FormControl variant="outlined" sx={{ flex: 1 }} size="small">
        <InputLabel data-testid="data-type-input" id="demo-simple-select-label">
          Data Type
        </InputLabel>
        <Select
          labelId="demo-simple-select-label"
          id="demo-simple-select"
          value={dataTypeValue}
          label="Data Type"
          onChange={handleTypeChange}
          data-testid="data-selector-field"
        >
          {renderMenuItems()}
        </Select>
      </FormControl>
      <TextField
        data-testid="name-input"
        id="outlined-basic"
        label="Field Name"
        variant="outlined"
        sx={{ flex: 1 }}
        onChange={handleNameChange}
        value={nameValue}
        size="small"
      />
      <IconButton
        aria-label="delete"
        size="small"
        onClick={handleDelete}
        data-testid="delete-button"
      >
        <DeleteIcon fontSize="inherit" />
      </IconButton>
    </div>
  );
};

export default DataSelectorField;
