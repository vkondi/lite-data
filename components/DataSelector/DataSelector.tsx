import { type FC, useCallback, useMemo } from "react";
import Button from "@mui/material/Button";
import AddIcon from "@mui/icons-material/Add";

import Layout from "../Layout/Layout";
import DataSelectorField from "./DataSelectorField";
import DataSelectorHeader from "./DataSelectorHeader";

import styles from "./DataSelector.module.css";
import {
  type FieldData,
  useDataSelectorContext,
} from "../../context/DataSelectorContext";

const emptyField: FieldData = { dataType: "", name: "" };

const DataSelector: FC = () => {
  const { fields, setFields } = useDataSelectorContext();

  const onTypeChange = useCallback((value: string, index: number) => {
    setFields((prevFields) =>
      prevFields.map((field, i) =>
        i === index ? { ...field, dataType: value, name: value } : field
      )
    );
  }, []);

  const onNameChange = useCallback((value: string, index: number) => {
    setFields((prevFields) =>
      prevFields.map((field, i) =>
        i === index ? { ...field, name: value } : field
      )
    );
  }, []);

  const onDelete = useCallback((index: number) => {
    setFields((prevFields) => prevFields.filter((_, i) => i !== index));
  }, []);

  const handleAddNewRow = () => {
    setFields((prevFields) => [...prevFields, emptyField]);
  };

  const renderedFields = useMemo(() => {
    return fields?.map((field, index) => (
      <DataSelectorField
        key={index}
        index={index}
        onTypeChange={onTypeChange}
        onNameChange={onNameChange}
        onDelete={onDelete}
        dataTypeValue={field.dataType}
        nameValue={field.name}
      />
    ));
  }, [fields, onTypeChange, onNameChange, onDelete]);

  return (
    <div className={styles.container}>
      <Layout>
        <div className={styles.layoutContainer}>
          <DataSelectorHeader />
          {renderedFields}
          <Button
            data-testid="add-new-row-button"
            variant="contained"
            startIcon={<AddIcon />}
            sx={{ width: 180, margin: "20px 0px 10px 0px" }}
            onClick={handleAddNewRow}
          >
            Add New Row
          </Button>
        </div>
      </Layout>
    </div>
  );
};

export default DataSelector;
