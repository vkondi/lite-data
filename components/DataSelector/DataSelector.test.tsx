import { render } from "@testing-library/react";
import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import DataSelector from "./DataSelector";
import { useDataSelectorContext } from '../../context/DataSelectorContext';

// Mock the context
vi.mock('../../context/DataSelectorContext', () => ({
  useDataSelectorContext: vi.fn()
}));

describe("DataSelector", () => {
  const mockFields = [{ field1: 'value1' }];

  beforeEach(() => {
    (useDataSelectorContext as any).mockReturnValue({
      fields: mockFields
    });
  });

  afterEach(() => {
    vi.clearAllMocks();
  });
  
  it("renders without crashing", () => {
    const { container } = render(<DataSelector />);
    expect(container).toMatchSnapshot();
  });
});
