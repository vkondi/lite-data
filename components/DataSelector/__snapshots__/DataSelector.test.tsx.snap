// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DataSelector > renders without crashing 1`] = `
<div>
  <div
    class="_container_61e5d5"
  >
    <div
      class="layout"
    >
      <div
        class="layout-content"
      >
        <div
          class="_layoutContainer_61e5d5"
        >
          <div
            class="_container_66254c"
          >
            <div
              class="MuiFormControl-root css-vav2vq-MuiFormControl-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
              >
                Data Type
              </label>
            </div>
            <div
              class="MuiFormControl-root css-vav2vq-MuiFormControl-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
              >
                Field Name
              </label>
            </div>
            <div
              class="_deleteIconSpacer_66254c"
              data-testid="delete-icon-spacer"
            />
          </div>
          <div
            class="_container_27718c"
          >
            <div
              class="MuiFormControl-root css-8p96bl-MuiFormControl-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1kfabtt-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                data-testid="data-type-input"
                id="demo-simple-select-label"
              >
                Data Type
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiSelect-root css-sc8y68-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
                data-testid="data-selector-field"
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-labelledby="demo-simple-select-label demo-simple-select"
                  class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-si86to-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                  id="demo-simple-select"
                  role="combobox"
                  tabindex="0"
                >
                  <span
                    aria-hidden="true"
                    class="notranslate"
                  >
                    â€‹
                  </span>
                </div>
                <input
                  aria-hidden="true"
                  aria-invalid="false"
                  class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
                  tabindex="-1"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-lohd6h-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-81qg8w"
                  >
                    <span>
                      Data Type
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root css-1lx7w30-MuiFormControl-root-MuiTextField-root"
              data-testid="name-input"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1kfabtt-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="outlined-basic"
                id="outlined-basic-label"
              >
                Field Name
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1pzfmz2-MuiInputBase-input-MuiOutlinedInput-input"
                  id="outlined-basic"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-81qg8w"
                  >
                    <span>
                      Field Name
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <button
              aria-label="delete"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-155wbr6-MuiButtonBase-root-MuiIconButton-root"
              data-testid="delete-button"
              id=":r2:"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-loadingIndicator css-165vu30-MuiIconButton-loadingIndicator"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1ckov0h-MuiSvgIcon-root"
                data-testid="DeleteIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
                />
              </svg>
            </button>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-2souiu-MuiButtonBase-root-MuiButton-root"
            data-testid="add-new-row-button"
            id=":r3:"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-1sh91j5-MuiButton-startIcon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="AddIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6z"
                />
              </svg>
            </span>
            Add New Row
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
